internal/grid/manager.go

```diff
@@ -0,0 +1,320 @@
+			id:           m.ID,
+			local:        o.Local,
+			remote:       host,
+			dial:         o.Dialer,
+			handlers:     &m.handlers,
+			auth:         o.AddAuth,
+			blockConnect: o.BlockConnect,
+			tlsConfig:    o.TLSConfig,
+			publisher:    o.TraceTo,
+		})
+	}
+	if !found {
+		return nil, fmt.Errorf("grid: local host not found")
+	}
+
+	return m, nil
+}
+
+// AddToMux will add the grid manager to the given mux.
+func (m *Manager) AddToMux(router *mux.Router) {
+	router.Handle(RoutePath, m.Handler())
+}
+
+// Handler returns a handler that can be used to serve grid requests.
+// This should be connected on RoutePath to the main server.
+func (m *Manager) Handler() http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		defer func() {
+			if debugPrint {
+				fmt.Printf("grid: Handler returning from: %v %v\n", req.Method, req.URL)
+			}
+			if r := recover(); r != nil {
+				debug.PrintStack()
+				err := fmt.Errorf("grid: panic: %v\n", r)
+				logger.LogIf(context.Background(), err, err.Error())
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+		}()
+		if debugPrint {
+			fmt.Printf("grid: Got a %s request for: %v\n", req.Method, req.URL)
+		}
+		ctx := req.Context()
+		if err := m.authRequest(req); err != nil {
+			logger.LogOnceIf(ctx, fmt.Errorf("auth %s: %w", req.RemoteAddr, err), req.RemoteAddr+err.Error())
+			w.WriteHeader(http.StatusForbidden)
+			return
+		}
+		conn, _, _, err := ws.UpgradeHTTP(req, w)
+		if err != nil {
+			if debugPrint {
+				fmt.Printf("grid: Unable to upgrade: %v. http.ResponseWriter is type %T\n", err, w)
+			}
+			w.WriteHeader(http.StatusUpgradeRequired)
+			return
+		}
+		defer conn.Close()
+		if debugPrint {
+			fmt.Printf("grid: Upgraded request: %v\n", req.URL)
+		}
+
+		msg, _, err := wsutil.ReadClientData(conn)
+		if err != nil {
+			logger.LogIf(ctx, fmt.Errorf("grid: reading connect: %w", err))
+			return
+		}
+		if debugPrint {
+			fmt.Printf("%s handler: Got message, length %v\n", m.local, len(msg))
+		}
+
+		var message message
+		_, _, err = message.parse(msg)
+		if err != nil {
+			if debugPrint {
+				fmt.Println("parse err:", err)
+			}
+			logger.LogIf(ctx, fmt.Errorf("handleMessages: parsing connect: %w", err))
+			return
+		}
+		if message.Op != OpConnect {
+			if debugPrint {
+				fmt.Println("op err:", message.Op)
+			}
+			logger.LogIf(ctx, fmt.Errorf("handler: unexpected op: %v", message.Op))
+			return
+		}
+		var cReq connectReq
+		_, err = cReq.UnmarshalMsg(message.Payload)
+		if err != nil {
+			if debugPrint {
+				fmt.Println("handler: creq err:", err)
+			}
+			logger.LogIf(ctx, fmt.Errorf("handleMessages: parsing ConnectReq: %w", err))
+			return
+		}
+		remote := m.targets[cReq.Host]
+		if remote == nil {
+			if debugPrint {
+				fmt.Printf("%s: handler: unknown host: %v. Have %v\n", m.local, cReq.Host, m.targets)
+			}
+			logger.LogIf(ctx, fmt.Errorf("handler: unknown host: %v", cReq.Host))
+			return
+		}
+		if debugPrint {
```

Comment:
Just a suggestion, should we have something like `DebugPrintIf` declared and used. will become one liner everywhere.

Reply:
No. This will be removed by the compiler since `debugPrint` is static when compiled. That means that objects that could escape through the print operation no longer does. 

Therefore when disabled it is a truly 0-cost function. If the compiler gets fixed to recognize this at some point we can do that.